File: create/pipeline/srs_generator.go
package pipeline

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"sync"

	"github.com/nyamiiko/script_hub/go_scripts/hub"
)

var extraSourceFiles = []string{
	filepath.Join(hub.ADBLOCK_DIR, "reject-drop.list"),
	filepath.Join(hub.ADBLOCK_DIR, "reject-no-drop.list"),
	filepath.Join(hub.CUSTOM_DIR, "substore.list"),
}

type SRSGenerator struct {
	singboxPath string
}

func NewSRSGenerator() *SRSGenerator {
	g := &SRSGenerator{}
	g.singboxPath = g.findSingbox()
	hub.EnsureDir(hub.SINGBOX_DIR)
	hub.EnsureDir(hub.CACHE_DIR)
	return g
}

func (g *SRSGenerator) findSingbox() string {
	out, err := exec.Command("which", "sing-box").Output()
	if err == nil && len(strings.TrimSpace(string(out))) > 0 {
		return strings.TrimSpace(string(out))
	}

	if hub.ValidateFileExists(hub.SINGBOX_LOCAL_BIN, "") {
		return hub.SINGBOX_LOCAL_BIN
	}
	if hub.ValidateFileExists(hub.SINGBOX_ROOT_BIN, "") {
		return hub.SINGBOX_ROOT_BIN
	}
	return ""
}

func (g *SRSGenerator) wildcardToRegex(value string) string {
	escaped := regexp.QuoteMeta(value)
	escaped = strings.ReplaceAll(escaped, `\*`, `.*`)
	escaped = strings.ReplaceAll(escaped, `\?`, `.`)
	return fmt.Sprintf("^%s$", escaped)
}

func (g *SRSGenerator) compileSrs(listFile string) {
	base := filepath.Base(listFile)
	name := strings.TrimSuffix(base, filepath.Ext(base))

	outDir := hub.SINGBOX_DIR
	if strings.Contains(listFile, "dns/mapping") || strings.Contains(listFile, "/dns/") {
		outDir = hub.SINGBOX_DNS_DIR
		hub.EnsureDir(outDir)
	} else if strings.Contains(listFile, "rulesets/AdBlock") {
		outDir = hub.ADBLOCK_DIR
		hub.EnsureDir(outDir)
	}

	srsFile := filepath.Join(outDir, fmt.Sprintf("%s_Singbox.srs", name))
	jsonTmp := filepath.Join(hub.CACHE_DIR, fmt.Sprintf("%s.json", name))
	jsonTmpWrite := jsonTmp + ".write"

	defer func() {
		os.Remove(jsonTmp)
		os.Remove(jsonTmpWrite)
	}()

	content := hub.ReadFileString(listFile)
	lines := strings.Split(content, "\n")

	mergedRules := map[string][]string{
		"domain":         {},
		"domain_suffix":  {},
		"domain_keyword": {},
		"domain_regex":   {},
		"ip_cidr":        {},
		"process_name":   {},
	}

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) < 2 {
			continue
		}
		rtype := strings.TrimSpace(parts[0])
		val := strings.TrimSpace(parts[1])

		switch rtype {
		case "DOMAIN":
			mergedRules["domain"] = append(mergedRules["domain"], val)
		case "DOMAIN-SUFFIX":
			mergedRules["domain_suffix"] = append(mergedRules["domain_suffix"], val)
		case "DOMAIN-KEYWORD":
			mergedRules["domain_keyword"] = append(mergedRules["domain_keyword"], val)
		case "DOMAIN-REGEX":
			mergedRules["domain_regex"] = append(mergedRules["domain_regex"], val)
		case "DOMAIN-WILDCARD":
			mergedRules["domain_regex"] = append(mergedRules["domain_regex"], g.wildcardToRegex(val))
		case "IP-CIDR", "IP-CIDR6":
			mergedRules["ip_cidr"] = append(mergedRules["ip_cidr"], val)
		case "PROCESS-NAME":
			mergedRules["process_name"] = append(mergedRules["process_name"], val)
		}
	}

	finalRules := make(map[string][]string)
	for k, v := range mergedRules {
		if len(v) > 0 {
			finalRules[k] = hub.UniqueStrings(v)
			sort.Strings(finalRules[k])
		}
	}

	srsJson := map[string]interface{}{
		"version": 1,
		"rules":   []interface{}{finalRules},
	}

	data, err := json.Marshal(srsJson)
	if err != nil {
		hub.Error(fmt.Sprintf("Error encoding JSON for %s: %v", name, err))
		return
	}

	if err := hub.SafeWriteFile(jsonTmpWrite, string(data), true); err != nil {
		hub.Error(fmt.Sprintf("Error writing JSON for %s: %v", name, err))
		return
	}
	os.Rename(jsonTmpWrite, jsonTmp)

	cmd := exec.Command(g.singboxPath, "rule-set", "compile", jsonTmp, "-o", srsFile)
	out, err := cmd.CombinedOutput()
	if err != nil {
		hub.Error(fmt.Sprintf("Failed to compile %s: %s", name, string(out)))
	} else {
		hub.Success(fmt.Sprintf("Compiled SRS: %s", name))
	}
}

func (g *SRSGenerator) pruneStaleSrs(listFiles []string) {
	expectedSrs := make(map[string]bool)
	for _, lf := range listFiles {
		base := filepath.Base(lf)
		name := strings.TrimSuffix(base, filepath.Ext(base))
		expectedSrs[fmt.Sprintf("%s_Singbox.srs", name)] = true
	}

	entries, _ := os.ReadDir(hub.SINGBOX_DIR)
	for _, entry := range entries {
		if strings.HasPrefix(entry.Name(), "GeoIP_") && strings.HasSuffix(entry.Name(), ".srs") {
			expectedSrs[entry.Name()] = true
		}
	}

	for _, entry := range entries {
		if !strings.HasSuffix(entry.Name(), ".srs") {
			continue
		}
		if !expectedSrs[entry.Name()] {
			path := filepath.Join(hub.SINGBOX_DIR, entry.Name())
			if err := os.Remove(path); err == nil {
				hub.Warn(fmt.Sprintf("Pruned stale SRS: %s", entry.Name()))
			} else {
				hub.Error(fmt.Sprintf("Failed to prune stale SRS: %s", entry.Name()))
			}
		}
	}

	if hub.ValidateDirExists(hub.SINGBOX_DNS_DIR, "") {
		dnsEntries, _ := os.ReadDir(hub.SINGBOX_DNS_DIR)
		for _, entry := range dnsEntries {
			if !strings.HasSuffix(entry.Name(), ".srs") {
				continue
			}
			if !expectedSrs[entry.Name()] {
				path := filepath.Join(hub.SINGBOX_DNS_DIR, entry.Name())
				if err := os.Remove(path); err == nil {
					hub.Warn(fmt.Sprintf("Pruned stale DNS SRS: %s", entry.Name()))
				} else {
					hub.Error(fmt.Sprintf("Failed to prune stale DNS SRS: %s", entry.Name()))
				}
			}
		}
	}
}

func (g *SRSGenerator) Run() {
	hub.Section("Sing-box SRS Generation (English Interface)")
	if g.singboxPath == "" {
		hub.Error("sing-box binary not found. Skipping SRS generation.")
		return
	}

	sourceDirs := []string{hub.RULE_SET_DIR, hub.DNS_MAPPING_DIR, hub.ADBLOCK_DIR}
	var listFiles []string

	for _, dir := range sourceDirs {
		if !hub.ValidateDirExists(dir, "") {
			continue
		}
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, entry := range entries {
			if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".list") {
				listFiles = append(listFiles, filepath.Join(dir, entry.Name()))
			}
		}
	}

	for _, path := range extraSourceFiles {
		if hub.ValidateFileExists(path, "") {
			listFiles = append(listFiles, path)
		}
	}

	listFiles = hub.UniqueStrings(listFiles)

	// Compile in parallel using a bounded worker pool matching the CPU count to maximize efficiency
	numWorkers := runtime.NumCPU()
	if numWorkers < 1 {
		numWorkers = 1
	}
	sem := make(chan struct{}, numWorkers)
	var wg sync.WaitGroup

	for _, lf := range listFiles {
		wg.Add(1)
		sem <- struct{}{}
		go func(path string) {
			defer wg.Done()
			defer func() { <-sem }()
			g.compileSrs(path)
		}(lf)
	}
	wg.Wait()

	g.pruneStaleSrs(listFiles)
}

File: create/processor/src/srs_generator.rs
use regex::escape;
use serde::Serialize;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Serialize)]
struct SingboxRuleset {
    version: i32,
    rules: Vec<HashMap<String, Vec<String>>>,
}

fn wildcard_to_regex(value: &str) -> String {
    let escaped = escape(value);
    let r = escaped.replace(r"\*", ".*").replace(r"\?", ".");
    format!("^{}$", r)
}

fn compile_srs(list_file: &Path, root_dir: &Path, singbox_path: &str) -> bool {
    let file_name = match list_file.file_name().and_then(|f| f.to_str()) {
        Some(f) => f,
        None => return false,
    };
    let name = file_name.trim_end_matches(".list");

    let mut out_dir = root_dir.join("rulesets/SingBox");
    let path_str = list_file.to_string_lossy();
    if path_str.contains("dns/mapping") || path_str.contains("/dns/") {
        out_dir = root_dir.join("rulesets/Sources/dns/mapping");
    } else if path_str.contains("rulesets/AdBlock") {
        out_dir = root_dir.join("rulesets/AdBlock");
    }

    if !out_dir.exists() {
        if fs::create_dir_all(&out_dir).is_err() {
            return false;
        }
    }

    let srs_file = out_dir.join(format!("{}_Singbox.srs", name));
    let cache_dir = root_dir.join("rulesets/Sources/Links/.cache");
    if !cache_dir.exists() {
        if fs::create_dir_all(&cache_dir).is_err() {
            return false;
        }
    }

    let json_tmp = cache_dir.join(format!("{}.json", name));
    let json_tmp_write = cache_dir.join(format!("{}.json.write", name));

    let content = match fs::read_to_string(list_file) {
        Ok(c) => c,
        Err(_) => return false,
    };

    let mut merged_rules: HashMap<String, HashSet<String>> = HashMap::new();
    merged_rules.insert("domain".to_string(), HashSet::new());
    merged_rules.insert("domain_suffix".to_string(), HashSet::new());
    merged_rules.insert("domain_keyword".to_string(), HashSet::new());
    merged_rules.insert("domain_regex".to_string(), HashSet::new());
    merged_rules.insert("ip_cidr".to_string(), HashSet::new());
    merged_rules.insert("process_name".to_string(), HashSet::new());

    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let parts: Vec<&str> = trimmed.split(',').collect();
        if parts.len() < 2 {
            continue;
        }
        let rtype = parts[0].trim();
        let val = parts[1].trim().to_string();

        match rtype {
            "DOMAIN" => {
                if let Some(set) = merged_rules.get_mut("domain") {
                    set.insert(val);
                }
            }
            "DOMAIN-SUFFIX" => {
                if let Some(set) = merged_rules.get_mut("domain_suffix") {
                    set.insert(val);
                }
            }
            "DOMAIN-KEYWORD" => {
                if let Some(set) = merged_rules.get_mut("domain_keyword") {
                    set.insert(val);
                }
            }
            "DOMAIN-REGEX" => {
                if let Some(set) = merged_rules.get_mut("domain_regex") {
                    set.insert(val);
                }
            }
            "DOMAIN-WILDCARD" => {
                if let Some(set) = merged_rules.get_mut("domain_regex") {
                    set.insert(wildcard_to_regex(&val));
                }
            }
            "IP-CIDR" | "IP-CIDR6" => {
                if let Some(set) = merged_rules.get_mut("ip_cidr") {
                    set.insert(val);
                }
            }
            "PROCESS-NAME" => {
                if let Some(set) = merged_rules.get_mut("process_name") {
                    set.insert(val);
                }
            }
            _ => {}
        }
    }

    let mut final_rules: HashMap<String, Vec<String>> = HashMap::new();
    for (k, v) in merged_rules {
        if !v.is_empty() {
            let mut list: Vec<String> = v.into_iter().collect();
            list.sort();
            final_rules.insert(k, list);
        }
    }

    let srs_json = SingboxRuleset {
        version: 1,
        rules: vec![final_rules],
    };

    let data = match serde_json::to_string(&srs_json) {
        Ok(d) => d,
        Err(_) => return false,
    };

    if fs::write(&json_tmp_write, data).is_err() {
        return false;
    }
    if fs::rename(&json_tmp_write, &json_tmp).is_err() {
        let _ = fs::remove_file(&json_tmp_write);
        return false;
    }

    let output = Command::new(singbox_path)
        .arg("rule-set")
        .arg("compile")
        .arg(&json_tmp)
        .arg("-o")
        .arg(&srs_file)
        .output();

    let _ = fs::remove_file(&json_tmp);

    match output {
        Ok(out) => {
            if out.status.success() {
                println!("\x1b[0;32m[SUCCESS]\x1b[0m Compiled SRS: {}", name);
                true
            } else {
                eprintln!(
                    "\x1b[0;31m[ERROR]\x1b[0m Failed to compile {}: {}",
                    name,
                    String::from_utf8_lossy(&out.stderr)
                );
                false
            }
        }
        Err(e) => {
            eprintln!(
                "\x1b[0;31m[ERROR]\x1b[0m Failed to execute sing-box: {:?}",
                e
            );
            false
        }
    }
}

fn prune_stale_srs(list_files: &[PathBuf], root_dir: &Path) {
    let mut expected_srs: HashSet<String> = HashSet::new();
    for lf in list_files {
        if let Some(file_name) = lf.file_name().and_then(|f| f.to_str()) {
            let name = file_name.trim_end_matches(".list");
            expected_srs.insert(format!("{}_Singbox.srs", name));
        }
    }

    let singbox_dir = root_dir.join("rulesets/SingBox");
    if let Ok(entries) = fs::read_dir(&singbox_dir) {
        for entry in entries.flatten() {
            if let Ok(name) = entry.file_name().into_string() {
                if name.starts_with("GeoIP_") && name.ends_with(".srs") {
                    expected_srs.insert(name);
                }
            }
        }
    }

    // Prune SINGBOX_DIR
    if let Ok(entries) = fs::read_dir(&singbox_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                if let Ok(name) = entry.file_name().into_string() {
                    if name.ends_with(".srs") && !expected_srs.contains(&name) {
                        if fs::remove_file(&path).is_ok() {
                            println!("\x1b[0;33m[WARN]\x1b[0m Pruned stale SRS: {}", name);
                        } else {
                            eprintln!(
                                "\x1b[0;31m[ERROR]\x1b[0m Failed to prune stale SRS: {}",
                                name
                            );
                        }
                    }
                }
            }
        }
    }

    // Prune SINGBOX_DNS_DIR
    let dns_dir = root_dir.join("rulesets/Sources/dns/mapping");
    if dns_dir.exists() {
        if let Ok(entries) = fs::read_dir(&dns_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_file() {
                    if let Ok(name) = entry.file_name().into_string() {
                        if name.ends_with(".srs") && !expected_srs.contains(&name) {
                            if fs::remove_file(&path).is_ok() {
                                println!("\x1b[0;33m[WARN]\x1b[0m Pruned stale DNS SRS: {}", name);
                            } else {
                                eprintln!(
                                    "\x1b[0;31m[ERROR]\x1b[0m Failed to prune stale DNS SRS: {}",
                                    name
                                );
                            }
                        }
                    }
                }
            }
        }
    }
}

pub fn run_srs_generator(root_path: &str, singbox_path: &str) -> bool {
    let root_dir = Path::new(root_path);
    if singbox_path.is_empty() {
        eprintln!(
            "\x1b[0;31m[ERROR]\x1b[0m sing-box binary path is empty. Skipping SRS generation."
        );
        return false;
    }

    let source_dirs = vec![
        root_dir.join("rulesets/RULE-SET"),
        root_dir.join("rulesets/Sources/dns/mapping"),
        root_dir.join("rulesets/AdBlock"),
    ];

    let mut list_files: Vec<PathBuf> = Vec::new();

    for dir in source_dirs {
        if !dir.exists() {
            continue;
        }
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_file() {
                    if let Some(ext) = path.extension() {
                        if ext == "list" {
                            list_files.push(path);
                        }
                    }
                }
            }
        }
    }

    // Add extra files
    let extra_source_files = vec![
        root_dir.join("rulesets/AdBlock/reject-drop.list"),
        root_dir.join("rulesets/AdBlock/reject-no-drop.list"),
        root_dir.join("rulesets/Sources/custom/substore.list"),
    ];

    for path in extra_source_files {
        if path.exists() {
            list_files.push(path);
        }
    }

    // Deduplicate
    let mut unique_files = HashSet::new();
    let mut deduped_list = Vec::new();
    for f in list_files {
        if unique_files.insert(f.clone()) {
            deduped_list.push(f);
        }
    }

    let mut success = true;
    for lf in &deduped_list {
        success &= compile_srs(lf, root_dir, singbox_path);
    }

    prune_stale_srs(&deduped_list, root_dir);
    success
}
